Add readeach function for iterating streams

non-Jedi opened the pull request:
See the commit messages and the discussion in #36132 for justification/explanation. <del>I can break this up into 2 PRs if there's only consensus on merging half of this, but it's semi-related, so I'm submitting as a single PR first.</del>

non-Jedi committed:
Add eachof function for iterating while reading from io

This saves boilerplate versus a while !eof(io); read(io, T) loop. The
function signature is patterned after the read(io, T) function
signature.

Prompted by #36132 but does not close #36132

I am open to better/clearer names for this function.

JeffBezanson commented:
I think these are both basically good ideas, but I agree they should be 2 PRs.

Unfortunately with the name `skipuntil` the meaning of the predicate should be inverted, so it's not a drop-in replacement for `skipchars` (but it could still replace it eventually).

non-Jedi commented:
Ah right. In my head `skipchars` was using the predicate the opposite of how it actually is. I'll remove those commits while I try to figure out the best way of using the deprecation machinery for this and let this PR be about `eachof`. Will file a separate PR for `skipuntil` later.

non-Jedi committed:
Keep skipchars from allocating unnecessary String

Remove usage of readline from skipchars definition.

non-Jedi committed:
Add NEWS entry for eachof

non-Jedi committed:
Refactor to make use of eachof function

These are all the places in base/stdlib where I saw the pattern used:

while!eof(io); read(io, T)

I think they make a good case study on this function's ability to make
code dealing with IO cleaner.

non-Jedi commented:
aarch64 failure seems unrelated

StefanKarpinski commented:
Can I suggest the name `readeach` instead of `eachof`?

non-Jedi commented:
`readeach` seems much better than `eachof`. I anticipated some discussion of what the best name for this was, so I'm going to wait to make the change to see if there are other suggestions.

non-Jedi commented:
Would `eachread` be better than `readeach` for symmetry with `eachline`, `eachindex`, etc.?

StefanKarpinski commented:
I think `readeach(io, T)` sounds more natural, personally, so I'd go with that.

non-Jedi committed:
Rename eachof to readeach

tkf commented on base/io.jl:
```diff
@@ -1003,6 +1000,46 @@ eltype(::Type{<:EachLine}) = String
 
 IteratorSize(::Type{<:EachLine}) = SizeUnknown()
 
+struct EachOfIO{T, IOT <: IO}
```
It'd make sense to rename `EachOfIO` too.

JeffBezanson replied:
Maybe `ReadEachIterator`?

non-Jedi replied:
Good catch. Thanks.

JeffBezanson commented:
Triage says :+1: to `readeach`.

non-Jedi committed:
Rename EachOfIO to ReadEachIterator

rfourquet commented on base/io.jl:
```diff
@@ -1003,6 +1000,46 @@ eltype(::Type{<:EachLine}) = String
+
+julia> open("my_file.txt") do io
+           for c in readeach(io, Char)
+               c == '\\n' && break
```
I'm confused: the doctests pass, but `'\\n'` doesn't seem like a valid character literal...

non-Jedi replied:
I believe that since this is within a string, the first backslash escapes the second.

rfourquet replied:
Oh right!

rfourquet commented on base/io.jl:
````diff
@@ -1003,6 +1000,46 @@ eltype(::Type{<:EachLine}) = String
+# Examples
+```jldoctest
+julia> open("my_file.txt", "w") do io
+           write(io, "JuliaLang is a GitHub organization.\\n It has many members.\\n");
````
For demonstration purposes, maybe using `IOBuffer("JuliaLang is a GitHub organization.\\n It has many members.\\n")` would be enough as the focus is on `readeach` ?

non-Jedi replied:
I copied the format of these doctests from the surrounding functions e.g. `readlines`. I'd prefer to keep them consistent, although I agree that using `IOBuffer` rather than opening an actual file might add to their clarity?

rfourquet commented on base/io.jl:
```diff
@@ -1087,10 +1124,10 @@ julia> String(readavailable(buf))
-        c = read(io, Char)
+    for c in readeach(io, Char)
         if c === linecomment
-            readline(io)
```
`readline(io)` is not valid anymore in this context? (sorry if this ovious!)

rfourquet replied:
Ok, I see from a commit message that this might be to avoid allocation which presumably happens in `readline`? you might add a comment to clarify this and prevent future "simplifications" changing this back to "readline" :)

StefanKarpinski reviewed (approved):
LGTM!

non-Jedi committed:
Revert substituting skipchars for readline

Although using skipchars reduced allocations, it overall slows down
the function.

non-Jedi committed:
Use IOBuffer for readeach example

This is more succinct and focused than writing a file to disk and
using open.

non-Jedi commented:
I've pushed two commits addressing @rfourquet's review that can be squashed once this is ready for merge.

tkf commented:
It looks like thre is a conflict in NEWS file.

mbauman committed:
Merge branch 'master' into eachof
